pkgs/build-support/build-fhs-userenv/chrootenv.c

```diff
@@ -20,56 +21,112 @@
-  char *env, **ret = malloc((len + 1) * sizeof(char *)), **ptr = ret;
-
-  for (size_t i = 0; i < len; i++) {
-    if ((env = getenv(names[i]))) {
-      if (asprintf(ptr++, "%s=%s", names[i], env) < 0)
-        errorf(EX_OSERR, "asprintf");
+#define LEN(x) sizeof(x) / sizeof(*x)
+
+char *env_blacklist[] = {};
+
+char **env_filter(char *envp[]) {
+  char **filtered_envp = malloc(sizeof(*envp));
+  size_t n = 0;
+
+  while (*envp != NULL) {
+    bool blacklisted = false;
+
+    for (size_t i = 0; i < LEN(env_blacklist); i++) {
+      if (!strncmp(*envp, env_blacklist[i], strlen(env_blacklist[i]))) {
+        blacklisted = true;
+        break;
+      }
+    }
+
+    if (!blacklisted) {
+      filtered_envp = realloc(filtered_envp, (n + 2) * sizeof(*envp));
+
+      if (filtered_envp == NULL)
+        errorf(EX_OSERR, "realloc");
+
+      filtered_envp[n++] = *envp;
     }
+
+    envp++;
+  }
+
+  filtered_envp[n] = NULL;
+  return filtered_envp;
+}
+
+void bind(char *from, char *to) {
+  if (mkdir(to, 0755) < 0)
+    errorf(EX_IOERR, "mkdir");
+
+  if (mount(from, to, "bind", MS_BIND | MS_REC, NULL) < 0)
+    errorf(EX_OSERR, "mount");
+}
+
+char *strjoin(char *dir, char *name) {
+  char *path = malloc(strlen(dir) + strlen(name) + 1);
+
+  if (path == NULL)
+    errorf(EX_OSERR, "malloc");
+
+  if (strcpy(path, dir) < 0)
+    errorf(EX_IOERR, "strcpy");
+
+  if (strcat(path, name) < 0)
+    errorf(EX_IOERR, "strcat");
+
+  return path;
+}
+
+char *bind_blacklist[] = {".", "..", "bin", "etc", "host", "usr"};
+
+bool bind_blacklisted(char *name) {
+  for (size_t i = 0; i < LEN(bind_blacklist); i++) {
+    if (!strcmp(bind_blacklist[i], name))
+      return true;
   }
 
-  *ptr = NULL;
-  return ret;
+  return false;
+}
+
+bool isdir(char *path) {
+  struct stat buf;
+  stat(path, &buf);
+  return S_ISDIR(buf.st_mode);
 }
 
-struct bind {
-  char *from;
-  char *to;
-};
+void bind_to_cwd(char *prefix) {
+  DIR *prefix_dir = opendir(prefix);
+
+  if (prefix_dir == NULL)
+    errorf(EX_OSERR, "opendir");
 
-struct bind binds[] = {{"/", "host"},   {"/proc", "proc"}, {"/sys", "sys"},
-                       {"/nix", "nix"}, {"/tmp", "tmp"},   {"/var", "var"},
-                       {"/run", "run"}, {"/dev", "dev"},   {"/home", "home"}};
+  struct dirent *prefix_dirent;
 
-void bind(struct bind *bind) {
-  DIR *src = opendir(bind->from);
+  while (prefix_dirent = readdir(prefix_dir)) {
+    if (bind_blacklisted(prefix_dirent->d_name))
+      continue;
 
-  if (src) {
-    if (closedir(src) < 0)
-      errorf(EX_IOERR, "closedir");
+    char *prefix_dirent_path = strjoin(prefix, prefix_dirent->d_name);
 
-    if (mkdir(bind->to, 0755) < 0)
-      errorf(EX_IOERR, "mkdir");
+    if (isdir(prefix_dirent_path)) {
+      bind(prefix_dirent_path, prefix_dirent->d_name);
+    } else {
+      char *host_target = strjoin("host/", prefix_dirent->d_name);
+
+      if (symlink(host_target, prefix_dirent->d_name) < 0)
+        errorf(EX_IOERR, "symlink");
```

Comment:
Directory is not closed and memory is not freed in error case.

Reply:
I don't understand why this has to be done. I'll read up on it.

Reply:
I think this should be handled by exit, not manually. Directories, files and memory are handled by default, if there's something that is not handled, `atexit` callback should be used.